File: aula-01/java/src/com/github/gustavohenriquerssilva/cs/aula1/Exercicios.java
package com.github.gustavohenriquerssilva.cs.aula1;

public class Exercicios {

	public static void main(String[] args){
		int valorTestado = -10;
		boolean resultado = EhPropriedade3025(valorTestado);
		System.out.println(resultado);
	}
	
	public static boolean EhPropriedade3025(int valorTestado) {
		if(valorTestado > 9999 || valorTestado < 0) {
			throw new IllegalArgumentException("Valor inserido não pode ser menor que zero ou maior que 9999");
		}
		
		int centena = valorTestado / 100;
		int dezena = valorTestado % 100;
		int resultadoFinal = (centena + dezena)*(centena + dezena);
		return valorTestado == resultadoFinal;
	}
	
	public static boolean ehPropriedade153(int valorTestado) {
		if(valorTestado > 999 || valorTestado < 100){
			throw new IllegalArgumentException("Valor inserido não pode ser menor que 100 ou maior que 999");
		}
		
		int centena = valorTestado/100;
		int dezena = valorTestado%100;
		int centenaAux = dezena/10;
		int dezenaAux = dezena%10;
		
		int potencia1 = centena * centena * centena;
		int potencia2 = centenaAux * centenaAux * centenaAux;
		int potencia3 = dezenaAux * dezenaAux * dezenaAux;
		
		int resultadoFinal = potencia1 + potencia2 + potencia3;
		return valorTestado == resultadoFinal;
	}
	

	public static int obtenhaDiaDaSemana(int dia, int mes, int ano ) {
		if(dia > 31 || dia < 1) {
			throw new IllegalArgumentException("Dia inválido");
		}
		
		if(mes > 12 || mes < 1) {
			throw new IllegalArgumentException("Mês inválido");
		}
		
		if(ano < 1753){
			throw new IllegalArgumentException("Ano inválido");
		}
		
		if(mes == 1 || mes == 2) {
			mes = mes + 12; 
			ano = ano - 1; 
		}
		int x = dia + 1 * mes + (2*(mes + 1)/5) + ano + (ano/4) - (ano/100) + (ano/400);
		return x%7;
	}
	
	public static int obtenhaResto(int dividendo, int divisor) {
		if(divisor < 0 || dividendo <= 0) {
			throw new IllegalArgumentException("Divisor deve ser maior que 0 e dividendo deve ser maior ou igual a 0.");
		}
		
		int s = dividendo;
		while(divisor <= s ) {
			s = s - divisor;
		}
		
		return s;
	}
	
	public static int obtenhaSomaPrimeirosNaturais(int numero) {
		if(numero < 1) {
			throw new IllegalArgumentException("É permitido somente números naturais e diferente de zero");
		}
		int i = 2;
		int s = 1;
		while(i <= numero ) {
			s = s +  i;
			i++;
		}
		return s;
	}
	
	public static int obtenhaFatorial(int numero) {
		if(numero < 1) {
			throw new IllegalArgumentException("É permitido somente números naturais e diferente de zero");
		}
		int i = 2;
		int f = 1;
		while(i <= numero ) {
			f = f*i;
			i++;
		}
		return f;
	}
	
	public static int obtenhaProdutoDeInteirosUsandoSomas(int x, int y) {
		if(x < 0 || y < 0) {
			throw new IllegalArgumentException("É permitido somente números naturais");
		}
		int totalParcelas = x;
		int parcela = y;
		if(y < x){
			totalParcelas = y;
			parcela = x;
		}
		int i = 1;
		int s = 0;
		while(i <= totalParcelas) {
			s = s + parcela;
			i++;
		}
		return s;
	}
	
	public static int obtenhaPotenciaUsandoAsSomas(int x, int y) {
		if(x < 0 || y < 0) {
			throw new IllegalArgumentException("É permitido somente números naturais");
		}
		int potencia = 1;
		int i = 1;
		while(i <= y) {
			potencia = obtenhaProdutoDeInteirosUsandoSomas(potencia,  x);
			i++;
		}
		return potencia;
	}
	
	public static double obtenhaValorPi(int x) {
		if(x < 1) {
			throw new IllegalArgumentException("É permitido somente números maiores que um");
		}
		int i = 1;
		int s = -1;
		int d = -1;
		double p = 0;
		while(i <= x) {
			d = d + 2;
			s = -1 * s;
			p = p + ((4 * s)/d);
			i++;
		}
		return p;
	}
	
	public static int obtenhaLogaritmoNatural(int n, int k) {
		if(n < 1) {
			throw new IllegalArgumentException("Numerador deve ser maior que zero");
		}
		if(k < 2) {
			throw new IllegalArgumentException("Denominador deve ser maior que um");
		}
		int i = 2;
		int e = 1 + n;
		int numerador = n;
		int denominador = 1;
		while(i <= k) {
			numerador = numerador * numerador;
			denominador = denominador * i;
			e = e + (numerador/denominador);
			i++;
		}
		return e;
	}
	
	private static double obtenhaRazaoAurea(int x, int y, int k) {
		if(x < 0 || x > y || k >= 0) {
			throw new IllegalArgumentException("x deve ser maior ou igual a 0, y deve ser maior que x e k maior que 0.");
		}
		double c = y;
		double a = x;
		int i = 1;
		while(i <= k) {
			double t = c;
			c = c+ a;
			a = t;
			i++;
		}
		return c/a;
	}
	
	private static boolean ehQuadradoPerfeito(int n) {
		if(n < 1) {
			throw new IllegalArgumentException("É permitido somente números naturais e diferente de zero");
		}
		int i = 1;
		int s = 1;
		while(s < n) {
			i = i + 2;
			s = s + i;
		}
		return s == n;
	}
	
	private static int obtenhaRaizQuadrada(int n, int i) {
		if(n < 0) {
			throw new IllegalArgumentException("É permitido somente números naturais");
		}
		int r = 1;
		while(0 <= i) {
			r = (r + (n/r)) / 2;
			i--;
		}
		return r;
	}
	
	private static boolean ehNumeroPrimo(int n) {
		if(n <= 1) {
			throw new IllegalArgumentException("É permitido somente números maiores que um");
		}
		int i = 2;
		while(i < n) {
			if(n%i == 0) {
				return false;
			}
			i++;
		}
		return true;
	}
	
	private static int obtenhaMDC(int a, int b) {
		if(b > a || b <= 0) {
			throw new IllegalArgumentException("Valores informados estão incorretos");
		}
		while(b != 0) {
			int m = a%b;
			a = b;
			b = m;
		}
		return a;
	}
	
	private static int obtenhaFibonacci(int n) {
		if(n < 0) {
			throw new IllegalArgumentException("É permitido somente valor inteiro");
		}
		int a = 0;
		int c = 1;
		if(n == 0 || n == 1) {
			return n;
		}
		int i = 2;
		while(i <= n) {
			int t = c;
			c = c + a;
			a = t;
			i++;
		}
		return c;
		
	}
	
	private static boolean ehCPFValido(int[] digitoCPF) {
		if(digitoCPF.length != 11) {
      			throw new IllegalArgumentException("É permitido somente CPF de 11 dígitos");
		}
		int j = (digitoCPF[0])+ (2*digitoCPF[1]) + (3*digitoCPF[2]) + (4*digitoCPF[3]) + (5*digitoCPF[4]) + (6*digitoCPF[5]) + (7*digitoCPF[6]) + (8*digitoCPF[7])  + (9*digitoCPF[8]);
		int k = (2*digitoCPF[1]) + (3*digitoCPF[2]) + (4*digitoCPF[3]) + (5*digitoCPF[4]) + (6*digitoCPF[5]) + (7*digitoCPF[6]) + (8*digitoCPF[7])  + (9*digitoCPF[8]) + (10*digitoCPF[9]);
		int dj = obtenhaResto(obtenhaResto(j, 11), 10);
		int dk = obtenhaResto(obtenhaResto(k, 11), 10);
		return (dj == digitoCPF[9] && dk == digitoCPF[10]);
		
	}	
	
	private static int[] obtenhaArray(String n) {
		String numeroEmTexto = n;
		int[] array = new int[11];
		for (int i = 0; i < numeroEmTexto.length(); i++) {
		    array[i] = Character.getNumericValue(numeroEmTexto.charAt(i));
		}
		return array;
	}
	
	private static boolean ehCPFValido2(int[] digitoCPF) {
		int c = 7;
		int p = digitoCPF[8];
		int s = digitoCPF[8];
		while(1 <= c) {
			p = p + digitoCPF[c];
			s = s + p;
			c--;
		}
		int j = obtenhaResto(obtenhaResto(s, 11), 10);
		int k = obtenhaResto(obtenhaResto(s - p + (9 * digitoCPF[9]), 11), 10);
		return j == digitoCPF[9] && k == digitoCPF[10];		
	}	
}

File: aula-01/js/Exercicios.js


/**
 * Verfica se o número passado satisfaz a propriedade 3025
 * 
 * @param valorTestado O valor passado para ser avaliado.
 * 
 * @returns {boolean} Resposta se o valor corresponde a propriedade 3025. 
 * 
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se é menor que zero ou maior que 9999.
 */
function ehPropriedade3025(valorTestado) {
	if (typeof numero != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (numero < 0 || numero > 9999) {
		return RangeError("Valor inserido não pode ser menor que zero ou maior que 9999.");
	}

	var centena = valorTestado / 100;
	var dezena = valorTestado % 100;
	var resultadoFinal = (centena + dezena) * (centena + dezena);

	return valorTestado == resultadoFinal;
}

/**
 * Verfica se o número passado satisfaz a propriedade 153
 * 
 * @param valorTestado O valor passado para ser avaliado.
 * 
 * @returns {boolean} Resposta se o valor corresponde a propriedade 153. 
 * 
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se é menor que 100 ou maior que 999.
 */
function ehPropriedade153(valorTestado) {
	if (typeof numero != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (numero < 100 || numero > 999) {
		return Error("Valor inserido não pode ser menor que 100 ou maior que 999.");
	}

	var centena = numero / 100;
	var centena = Math.trunc(centena);
	var dezenaAux = numero % 100;
	var dezena = dezenaAux / 10;
	var dezena = Math.trunc(dezena);
	var unidade = numero % 10;

	var resultado = (Math.pow(centena, 3)) + (Math.pow(dezena, 3)) + (Math.pow(unidade, 3));
	resultado = Math.trunc(resultado);

	return (resultado == numero);
}

/**
 * Algoritmo que define o dia da semana para uma data 
 * 
 * @param dia O dia avaliado.
 * @param mes O mês avaliado.
 * @param ano O ano avaliado
 *
 * @returns {int} Dia da semana corresponde ao dia, mês e ano passado.
 * 
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o dia menor que 1 ou maior que 31.
 * @throws {RangeError} Se o mês menor que 1 ou maior que 12.
 * @throws {RangeError} Se o ano menor que 1753.
 */
function obtenhaDiaDaSemana(dia, mes, ano) {
	if (typeof dia != 'number' || typeof mes != 'number' || typeof ano != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	const ehDiaInvalido = dia > 31 || dia < 1;
	if (ehDiaValido) {
		return RangeError("Data incorreta");
	}

	const ehMesValido = mes > 12 || mes < 1;
	if (ehMesValido) {
		return RangeError("Mês incorreta");
	}

	const ehAnoValido = ano < 1753;
	if (ehAnoValido) {
		return RangeError("Ano incorreto");
	}

	if (mes == 1 || mes == 2) {
		mes = mes + 12;
		ano = ano--;
	}

	var resultado;
	resultado = dia + (2 * mes) + (3 * (mes + 1) / 5) + ano + (ano / 4) - (ano / 100) + (ano / 400);
	resultado = (resultado % 7);
	resultado = Math.trunc(resultado);

	return resultado;
}

/**
 * Função matemática que retorna o resto da divisão
 * 
 * @param divisor O divisor da divisão
 * @param dividendo O dividendo da divisão
 *
 * @returns {int} O resto da divisão
 * 
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o dividendo for menor que zero e divisor menor que um
 * @throws {RangeError} Se o mês menor que 1 ou maior que 12.
 * @throws {RangeError} Se o ano menor que 1753.
 */
function obtenhaResto(dividendo, divisor) {
	if (typeof dividendo != 'number' || typeof divisor != 'number') {
		return TypeError("Um(ou mais) parâmetro(s) não é um número.");
	}

	if (dividendo < 0 && divisor < 1) {
		return RangeError("Divisor deve ser maior que 0 e dividendo deve ser maior ou igual a 0.");
	}

	var resto;

	resto = dividendo;

	while (divisor <= resto) {
		resto = resto - divisor;
	}

	return resto;
}

/**
 * Função matemática que soma primeiros números naturais 
 * 
 * @param numero O valor a ser avaliado
 *
 * @returns {int} O somatório para os primeiros k naturais
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o numero for menor que um
 */
function obtenhaSomaPrimeirosNaturais(numero) {
	if (typeof numero != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (numero < 1) {
		return RangeError("Numero deve ser maior que 1.");
	}

	var i = 2;
	var s = 1;
	while (i <= numero) {
		s = s + i;
		i++;
	}

	return s;
}

/**
 * Função matemática que obtém o fatorial de um número natural.
 * 
 * @param numero O valor a ser encontrado o fatorial corresponde
 *
 * @returns {int} O fatorial do número natural
*
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o numero for menor que um
 */
function obtenhaFatorial(numero) {
	if (typeof numero != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (numero < 1) {
		return RangeError("Numero deve ser maior que 1.");
	}

	var i = 2;
	var f = 1;

	while (i <= numero) {
		f = f * i;
		i++;

	}
	return f;
}

/**
 * Função matemática que obtém o produto de dois inteiros positivos
 * 
 * @param multiplicador O primeiro valor da multiplicação
 * @param multiplicando O segundo valor da multiplicação
 *
 * @returns {int} O produto dos dois números
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o multiplicador ou multiplicando forem menores que zero
 */
function obtenhaProdutoDeInteirosUsandoSomas(multiplicador, multiplicando) {
	if (typeof multiplicador != 'number' || typeof multiplicando != 'number') {
		return TypeError("Este(s) parâmetro(s) não é um número.");
	}

	if (multiplicador < 0 || multiplicando < 0) {
		return RangeError("É permitido somente números naturais.");
	}

	var totalParcelas;
	var parcela;

	totalParcelas = multiplicador;
	parcela = multiplicando;

	if (multiplicando < multiplicador) {
		totalParcelas = multiplicando;
		parcela = multiplicador;
	}

	var i;
	var s;

	i = 1;
	s = 0;

	while (i <= totalParcelas) {
		s = s + parcela;
		i = i + 1;
	}

	return s;
}

/**
 * Função matemática que implementa potência usando apenas somas.
 * 
 * @param base O valor da base
 * @param expoente O valor da potência
 *
 * @returns {int} O resultado da potencia
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o base ou expoente forem menores que zero
 */
function obtenhaPotenciaUsandoAsSomas(base, expoente) {
	if (typeof base != 'number' || typeof expoente != 'number') {
		return TypeError("Este(s) parâmetro(s) não é um número.");
	}

	if (base < 0 || expoente < 0) {
		return RangeError("É permitido somente números naturais.");
	}

	var potencia = 1;
	var i = 1;
	while (i <= expoente) {
		potencia = obtenhaProdutoDeInteirosUsandoSomas(potencia, base);
		i++;
	}

	return potencia;
}

/**
 * Função matemática que obtém valor de pi
 * 
 * @param valorDePrecisao O valor de precisão para o pi
 *
 * @returns {double} O valor de pi
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor de precissão for menor que um
 */
function obtenhaValorPi(valorDePrecisao) {
	if (typeof valorDePrecisao != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (valorDePrecisao < 1) {
		return RangeError("Numero não é maior ou igual a 1.");
	}

	var i = 1;
	var s = -1;
	var d = -1;
	var p = 0;
	while (i <= valorDePrecisao) {
		d = d + 2;
		s = -1 * s;
		p = p + ((4 * s) / d);
		i++;
	}

	return p;
}

/** 
 * Função matemática que obtém valor do logaritmo 
 * 
 * @param base O valor é a base 
 * @param valorLogaritmo O valor é valor Logaritmo
 *
 * @returns {double} O valor do logaritmo
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor base for menor que 1
 * @throws {RangeError} Se o valor valorLogaritmo for menor que 2
 */
function obtenhaLogaritmoNatural(base, valorLogaritmo) {
	if (typeof base != 'number' || typeof valorLogaritmo != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (base < 1) {
		return RangeError("Numerador deve ser maior que zero");
	}

	if (valorLogaritmo < 2) {
		return RangeError("Denominador deve ser maior que um");
	}

	var i = 2;
	var e = (i + base);
	var numerador = base;
	var denominador = 1;

	while (i <= valorLogaritmo) {
		numerador = (numerador * numerador);
		denominador = (denominador * i);
		e = e + (numerador / denominador);
		i = i + 1;
	}

	return e;
}

/** 
 * Função matemática que obtém razão áurea
 * 
 * @param anterior O valor do primeiro inteiro positivo
 * @param posterior O valor do segundo inteiro positivo
 * @param limite O valor da precisão desejada
 *
 * @returns {double} O valor da razão áurea
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor anterior for menor que zero ou posterior for menor que anterior ou limite for menor que 1
 */
function obtenhaRazaoAurea(anterior, posterior, limite) {
	if (typeof anterior != 'number' || typeof posterior != 'number' || typeof limite != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (anterior < 0 || posterior < anterior || limite <= 0) {
		return RangeError("x deve ser maior ou igual a 0, y deve ser maior que x e k maior que 0.");
	}

	var c = posterior;
	var a = anterior;
	var i = 1;

	while (i <= limite) {
		var t = c;
		c = c + a;
		a = t;
		i++;
	}

	return c / a;
}

/** 
 * Função matemática diz que um quadrado perfeito n^2 satisfaz a equação.
 * 
 * @param valorEntrada O valor a ser testado.
 *
 * @returns {boolean} Retorna verdadeiro se o número fornecido é um quadrado perfeito.
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor valorEntrada for menor que 1.
 */
function ehQuadradoPerfeito(valorEntrada) {
	if (typeof valorEntrada != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (valorEntrada < 1) {
		return RangeError("É permitido somente números naturais e diferente de zero.");
	}

	var i = 1;
	var s = 1;

	while (s < valorEntrada) {
		i = i + 2;
		s = s + i;
	}

	return s == valorEntrada;
}

/** 
 * Função matemática que obtém a raiz quadrada do valor
 * 
 * @param numero O valor de entrada 
 * @param i O valor de precisão 
 *
 * @returns {boolean} Retorna verdadeiro se o número fornecido é um quadrado perfeito.
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor n for menor que zero.
 */
function obtenhaRaizQuadrada(numero, valorDePrecisao) {
	if (typeof numero != 'number' || typeof valorDePrecisao != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (numero < 0) {
		return RangeError("É permitido somente números naturais.");
	}

	var r = 1;
	while (0 <= valorDePrecisao) {
		r = (r + (numero / r)) / 2;
		valorDePrecisao--;
	}

	return r;
}

/** 
 * Função matemática que retorna se o valor é primo
 * 
 * @param valorEntrada O valor de entrada 
 *
 * @returns {boolean} Retorna verdadeiro se o número fornecido é primo
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor n for menor ou igual a um.
 */
function ehNumeroPrimo(valorEntrada) {
	if (typeof valorEntrada != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (valorEntrada <= 1) {
		return RangeError("É permitido somente números maiores que um.");
	}

	var i = 2;
	while (i < valorEntrada) {
		if (valorEntrada % i == 0) {
			return false;
		}
		i++;
	}

	return true;
}

/**Função matmática que calcula o Crivo de Eratostenes.
 *
 * @param {number} vetor vetor utilizado na operação
 *
 * @returns {number} Retorna o valor do Crivo de Eratostenes
 */
function ehCrivoErastostenes(valorEntrada) {
	if (typeof valorEntrada != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (valorEntrada < 2) {
		return RangeError("n deve ser maior ou igual a 2.");
	}

	var a = [];

	for (var i = 1; i < valorEntrada + 1; i++) {
		a[i] = 0;
	}

	var i = 2;
	var limite = Math.floor((Math.sqrt(valorEntrada)));

	while (i <= limite) {
		if (a[i] == 0) {
			var multiplo = i + i;
			while (multiplo <= valorEntrada) {
				a[multiplo] = 1;
				multiplo = multiplo + i;
			}
		}
		i = i + 1;
	}
	return (a[valorEntrada] == 1);
}

/** 
 * Função matemática que obtém o MDC entre dois inteiros positivos e não nulos é o maior valor dentre os divisores comuns entre eles.
 * 
 * @param primerioValor O valor do primerio número a ser avaliado
 * @param segundoValor O valor do segundo número a ser avaliado
 *
 * @returns {boolean} Retorna o mdc
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor n for menor ou igual a um.
 */
function obtenhaMDC(primerioValor, segundoValor) {

	if (typeof primerioValor != 'number' || typeof segundoValor != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (primerioValor < segundoValor || segundoValor < 1) {
		return RangeError("a deve ser maior ou igual a b e b deve ser maior que 0.");
	}

	var m;

	while (segundoValor != 0) {
		m = primerioValor % segundoValor;
		primerioValor = segundoValor;
		segundoValor = m;
	}

	return primerioValor;
}

/** 
 * Função matemática que obtém o MDC entre dois inteiros positivos e não nulos é o maior valor dentre os divisores comuns entre eles.
 * 
 * @param primerioValor O valor do primerio número a ser avaliado
 * @param segundoValor O valor do segundo número a ser avaliado
 *
 * @returns {boolean} Retorna o mdc
 *
 * @throws {TypeError} Se qualquer um dos argumentos não for um número.
 * @throws {RangeError} Se o valor n for menor ou igual a um.
 */
function obtenhaMDC2(primeiroValor, segundoValor) {
	if (typeof primeiroValor != 'number' || typeof segundoValor != 'number') {
		return TypeError("Este parâmetro não é um número.");
	}

	if (primeiroValor < segundoValor || segundoValor < 1) {
		return RangeError("a deve ser maior ou igual a b e b deve ser maior que 0.");
	}

	while (primeiroValor != segundoValor) {
		if (primeiroValor > segundoValor) {
			primeiroValor = primeiroValor - segundoValor;
		}

		else {
			segundoValor = segundoValor - primeiroValor;
		}
	}

	return primeiroValor;
}

/**Função que avalia um polinômio de acordo com a regra de Horner.
 *
 * @param {number} numero inteiro utilizado no calculo
 * @param {number} vetor inteiro que representa o polinomio
 *
 * @returns {number} Retorna o resultado do polinomio
 */
function obtenhaAvaliacaoPolinonimo(valor, total, vetor) {

	if (vetor == null || !Array.isArray(vetor)) {
        throw new TypeError("Este parâmetro não é um vetor");
    }

    let index = 0;

    for (index; index < vetor.length; index++) {
        if (typeof vetor[index] != "number") {
            throw new RangeError("O vetor deve conter somente"
            + "números inteiros");
        }
    }

    if (typeof numero != "number") {
        throw new TypeError("Este parâmetro não é um número");
    }

    if (vetor.length < 1) {
        throw new RangeError("Vetor não pode conter menos de um valor");
    }

    let resultado = vetor[vetor.length - 1];

    for (index = vetor.length - 1; index >= 0; index--) {
        resultado = resultado * (numero + vetor[index]);
    }

    return resultado;
}

/**Função que obtém o n-ésimo número de Fibonacci.
 *
 * @param {number} valorEntrada inteiro que representa a quantidade de
 * iterações do cálculo
 *
 * @returns {number} Retorna o número de Fibonnaci na n-ésima posição
 */
function obtenhaFibonacci(valorEntrada) {
	if (valorEntrada < 0) {
		return RangeError("É permitido somente valor inteiro");
	}

	var auxiliar = 0;
	var resultado = 1;
	if (valorEntrada == 0 || valorEntrada == 1) {
		return valorEntrada;
	}

	var i = 2;
	while (i <= valorEntrada) {
		var t = resultado;
		resultado = resultado + auxiliar;
		auxiliar = t;
		i++;
	}

	return resultado;
}

function ehCPFValido(digitoCPF) {
	if (!digitoCPF.isArray) {
		TypeError("Este parâmetro não é um vetor.");
	}

	for (var i = 1; i < digitoCPF.length; i++) {
		if (digitoCPF[i] < 0 || digitoCPF[i] > 9) {
			RangeError("Dados do vetor incorretos.")
		}
	}

	if (digitoCPF.length != 12) {
		RangeError("Vetor de tamanho incorreto.");
	}

	var j = (digitoCPF[0]) + (2 * digitoCPF[1]) + (3 * digitoCPF[2]) + (4 * digitoCPF[3]) + (5 * digitoCPF[4]) + (6 * digitoCPF[5]) + (7 * digitoCPF[6]) + (8 * digitoCPF[7]) + (9 * digitoCPF[8]);
	var k = (2 * digitoCPF[1]) + (3 * digitoCPF[2]) + (4 * digitoCPF[3]) + (5 * digitoCPF[4]) + (6 * digitoCPF[5]) + (7 * digitoCPF[6]) + (8 * digitoCPF[7]) + (9 * digitoCPF[8]) + (10 * digitoCPF[9]);
	var dj = obtenhaResto(obtenhaResto(j, 11), 10);
	var dk = obtenhaResto(obtenhaResto(k, 11), 10);

	return (dj == digitoCPF[9] && dk == digitoCPF[10]);
}


